libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/route/NavigationRouteEngine.java

```diff
@@ -0,0 +1,195 @@
+
+  private static final double BEARING_TOLERANCE = 90d;
+
+  private List<RouteEngineListener> routeEngineListeners = new CopyOnWriteArrayList<>();
+
+  private String accessToken;
+  private String routeProfile;
+  private RouteProgress routeProgress;
+  private Locale locale = Locale.getDefault();
+  private int unitType = NavigationUnitType.NONE_SPECIFIED;
+
+  @Override
+  public void addRouteEngineListener(RouteEngineListener listener) {
+    if (!routeEngineListeners.contains(listener)) {
+      routeEngineListeners.add(listener);
+    }
+  }
+
+  @Override
+  public void removeRouteEngineListener(RouteEngineListener listener) {
+    routeEngineListeners.remove(listener);
+  }
+
+  @Override
+  public void updateAccessToken(String accessToken) {
+    this.accessToken = accessToken;
+  }
+
+  @Override
+  public void updateLocale(Locale locale) {
+    this.locale = locale;
+  }
+
+  @Override
+  public void updateUnitType(int unitType) {
+    this.unitType = unitType;
+  }
+
+  @Override
+  public void updateRouteProfile(String routeProfile) {
+    if (RouteUtils.isValidRouteProfile(routeProfile)) {
+      this.routeProfile = routeProfile;
+    }
+  }
+
+  /**
+   * Calculates a new {@link com.mapbox.api.directions.v5.models.DirectionsRoute} given
+   * the current {@link Location} and {@link RouteProgress} along the route.
+   * <p>
+   * Uses {@link RouteOptions#coordinates()} and {@link RouteProgress#remainingWaypoints()}
+   * to determine the amount of remaining waypoints there are along the given route.
+   *
+   * @param location      current location of the device
+   * @param routeProgress for remaining waypoints along the route
+   * @since 0.13.0
+   */
+  @Override
+  public void findRouteFromRouteProgress(Location location, RouteProgress routeProgress) {
+    if (isValidProgress(location, routeProgress)) {
+      return;
+    }
+    this.routeProgress = routeProgress;
+    Point origin = Point.fromLngLat(location.getLongitude(), location.getLatitude());
+    Double bearing = location.hasBearing() ? Float.valueOf(location.getBearing()).doubleValue() : null;
+    NavigationRoute.Builder builder = buildRouteRequestFromCurrentLocation(
+      origin, bearing, routeProgress, routeProfile
+    );
+    if (builder != null) {
+      builder.accessToken(accessToken);
+      addLocaleAndUnitType(builder);
+      builder.build().getRoute(directionsResponseCallback);
+    }
+  }
+
+  @Override
+  public void findRouteFromOriginToDestination(Point origin, Point destination) {
+    NavigationRoute.Builder builder = NavigationRoute.builder()
+      .accessToken(accessToken)
+      .origin(origin)
+      .destination(destination);
+    addLocaleAndUnitType(builder);
+    addRouteProfile(routeProfile, builder);
+    builder.build().getRoute(directionsResponseCallback);
+  }
+
+  @Nullable
+  private static NavigationRoute.Builder buildRouteRequestFromCurrentLocation(Point origin, Double bearing,
```

Comment:
Why `buildRouteRequestFromCurrentLocation` needs to be `static`? Same for `addRouteProfile`, `addDestination`, `retrieveDestinationWaypoint` and `addWaypoints` methods.